pkg/app/api/grpcapi/web_api.go

```diff
@@ -194,6 +197,83 @@ func (a *WebAPI) DeleteEnvironment(ctx context.Context, req *webservice.DeleteEn
 	if err := a.validateEnvBelongsToProject(ctx, req.EnvironmentId, claims.Role.ProjectId); err != nil {
 		return nil, err
 	}
+	// Check if no Piped has permission to the given environment.
+	pipeds, err := a.pipedStore.ListPipeds(ctx, datastore.ListOptions{
+		Filters: []datastore.ListFilter{
+			{
+				Field:    "EnvIds",
+				Operator: datastore.OperatorIn,
+				Value:    req.EnvironmentId,
+			},
+			{
+				Field:    "ProjectId",
+				Operator: datastore.OperatorEqual,
+				Value:    claims.Role.ProjectId,
+			},
+		},
+	})
+	if err != nil {
+		a.logger.Error("failed to fetch Pipeds linked to the given environment",
+			zap.String("env-id", req.EnvironmentId),
+			zap.Error(err),
+		)
+		return nil, status.Error(codes.Internal, "Failed to validate the deletion operation")
+	}
+	if len(pipeds) > 0 {
+		var pipedNames strings.Builder
+		for _, p := range pipeds {
+			pipedNames.WriteString(fmt.Sprintf("%q", p.Name))
```

Comment:
looks like this pipedNames output has no dividers in between, right? 🤔  I think just make an array of names and merge them using [strings.Join](https://golang.org/pkg/strings/#Join) could be simpler.

Reply:
Exactly, that's more suitable for this 👍